### ISSUE TITLE
Source is not cleaned when breakfast fails

### ISSUE THREAD

**Issue by petefoth** [lineageos4microg/docker-lineage-cicd#599]
If a call to `breakfast` fails, the [call to `continue`]( https://github.com/lineageos4microg/docker-lineage-cicd/blob/cfc5745d04d5a0300c96b0a0968490b3373cf7b7/src/build.sh#L408) bypasses the cleanup code in [lines 523-L547](https://github.com/lineageos4microg/docker-lineage-cicd/blob/cfc5745d04d5a0300c96b0a0968490b3373cf7b7/src/build.sh#L523-L547)

Solution is to put the cleanup code in functions and call the functions before continuing.

Note this may be the root cause of #582

**Comment by petefoth** [comment]
*Wed 03 Apr 24*
- Pushed a possible fix to `pf-functionise-cleanup` branch
- test in a docker instance: 
    - build for `foster` (which should fail because it's a TV box) the `FP3` (which should succeed, but failed in the April build run because kernel code from `foster` build was still present)
    - set "BUILD_OVERLAY=true" to see if this fixes #582
```
docker run \
    --name "lineageos4microg-$(date +%Y%m%d%I%M)" \
    --cap-add=SYS_ADMIN \
    --security-opt apparmor=unconfined \
    --security-opt seccomp=unconfined \
    -e "BRANCH_NAME=lineage-20.0" \
    -e "USE_CCACHE=true" \
    -e "CCACHE_SIZE=500G" \
    -e "RELEASE_TYPE=microG" \
    -e "OTA_URL=https://download.lineage.microg.org/api" \
    -e "INCLUDE_PROPRIETARY=true" \
    -e "CLEAN_OUTDIR=false" \
    -e "CLEAN_AFTER_BUILD=true" \
    -e "BUILD_OVERLAY=true" \
    -e "SIGN_BUILDS=true" \
    -e "ZIP_SUBDIR=true" \
    -e "LOGS_SUBDIR=true" \
    -e "SIGNATURE_SPOOFING=no" \
    -e "DELETE_OLD_ZIPS=2" \
    -e "DELETE_OLD_LOGS=2" \
    -e "WITH_GMS=true" \
    -v "/srv/src:/srv/src" \
    -v "/srv/tmp:/srv/tmp" \
    -v "/srv/ccache:/srv/ccache" \
    -v "/srv/mirror:/srv/mirror" \
    -v "/srv/zips:/srv/zips" \
    -v "/srv/logs:/srv/logs" \
    -v "/srv/keys:/srv/keys" \
    -v "/srv/local_manifests:/srv/local_manifests" \
    -v "/srv/userscripts:/srv/userscripts" \
    -e "DEVICE_LIST=foster,FP3" \
    lineageos4microg/docker-lineage-cicd:pf-functionise-cleanup
```
- that got past the point where the bulid failed
- try with the `latest`, `master` image, and without the `-e "BUILD_OVERLAY=true" \`
```
docker run \
    --name "lineageos4microg-$(date +%Y%m%d%I%M)" \
    --cap-add=SYS_ADMIN \
    --security-opt apparmor=unconfined \
    --security-opt seccomp=unconfined \
    -e "BRANCH_NAME=lineage-20.0" \
    -e "USE_CCACHE=true" \
    -e "CCACHE_SIZE=500G" \
    -e "RELEASE_TYPE=microG" \
    -e "OTA_URL=https://download.lineage.microg.org/api" \
    -e "INCLUDE_PROPRIETARY=true" \
    -e "CLEAN_OUTDIR=false" \
    -e "CLEAN_AFTER_BUILD=true" \
    -e "SIGN_BUILDS=true" \
    -e "ZIP_SUBDIR=true" \
    -e "LOGS_SUBDIR=true" \
    -e "SIGNATURE_SPOOFING=no" \
    -e "DELETE_OLD_ZIPS=2" \
    -e "DELETE_OLD_LOGS=2" \
    -e "WITH_GMS=true" \
    -v "/srv/src:/srv/src" \
    -v "/srv/tmp:/srv/tmp" \
    -v "/srv/ccache:/srv/ccache" \
    -v "/srv/mirror:/srv/mirror" \
    -v "/srv/zips:/srv/zips" \
    -v "/srv/logs:/srv/logs" \
    -v "/srv/keys:/srv/keys" \
    -v "/srv/local_manifests:/srv/local_manifests" \
    -v "/srv/userscripts:/srv/userscripts" \
    -e "DEVICE_LIST=foster,FP3" \
    lineageos4microg/docker-lineage-cicd
```
- that failed as expected

**Comment by petefoth** [comment]
Fixed by https://github.com/lineageos4microg/docker-lineage-cicd/pull/600
